feat: config to enable resilient registry (#2440)
defenseunicorns/zarf: site/src/content/docs/ref/init-package.mdx

:::

> ## Making the Registry Highly-Availabile

Reviewer: ```suggestion
## Making the Registry Highly-Available
```

> By default the registry included in the init package creates a ReadWriteOnce PVC and only schedules onto one node at a time. For smaller and simpler deployments this is usually enough. However, for larger deployments or deployments where nodes will be restarted or updated somewhat frequently, you may want to make the registry highly-available.

Reviewer: ```suggestion
By default, the registry included in the init package creates a ReadWriteOnce PVC and is only scheduled to run on one node at a time. This setup is usually enough for smaller and simpler deployments. However, for larger deployments or those where nodes are frequently restarted or updated, you may want to make the registry highly-available.
```

> This can have some prerequisites, such as having a storage class that supports ReadWriteMany, or being in an environment where you can configure the registry to use an s3 compatible backend. Beyond the backing storage, though, you will need to pass some custom configuration to the registry so that it spreads out across all nodes and has the correct amount of replicas. Below is an example configuration using a ReadWriteMany storage class:

Reviewer: ```suggestion
This approach requires certain prerequisites, such as a storage class that supports ReadWriteMany, or being in an environment that allows you to configure the registry to use an S3-compatible backend. Additionally, you must provide custom configuration to the registry to ensure it is distributed across all nodes and has the appropriate number of replicas. Below is an example configuration using a ReadWriteMany storage class:
```

> This approach requires certain prerequisites, such as a storage class that supports ReadWriteMany, or being in an environment that allows you to configure the registry to use an S3-compatible backend. Additionally, you must provide custom configuration to the registry to ensure it is distributed across all nodes and has the appropriate number of replicas. Below is an example configuration using a ReadWriteMany storage class:
> 
> ```yaml
> package:
>   deploy:

Reviewer 2: ```suggestion
This approach requires certain prerequisites, such as a storage class that supports ReadWriteMany, or being in an environment that allows you to configure the registry to use an S3-compatible backend. Additionally, you must provide custom configuration to the registry to ensure it is distributed across all nodes and has the appropriate number of replicas. Below is an example [configuration file](/ref/config-files/) using a ReadWriteMany storage class:

```yaml
# zarf-config.yaml
package:
```

>     set:

Reviewer: I think we will need `deploy` in here

https://github.com/defenseunicorns/zarf/blob/b4201718aa96f34f37dac7afa5ccfe59affedbd0/examples/config-file/zarf-config.yaml#L10-L13

Reviewer 2: thats mb, looks like my `suggest` diff didnt sync properly

      REGISTRY_PVC_ENABLED: "true"
      REGISTRY_PVC_ACCESS_MODE: "ReadWriteMany"
      REGISTRY_HPA_AUTO_SIZE: "true"
      REGISTRY_AFFINITY_CUSTOM: |
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchExpressions:
                    - key: app
                      operator: In
                      values:
                        - docker-registry
                topologyKey: kubernetes.io/hostname
```

> Of note is the `REGISTRY_AFFINITY_CUSTOM` variable which overrides the default pod anti-affinity and `REGISTRY_HPA_AUTO_SIZE` which will automatically set the min and max replicas for the registry based on the number of nodes in the cluster. If you would prefer to set the min and max replicas manually you can set `REGISTRY_HPA_MIN` and `REGISTRY_HPA_MAX` to the desired values instead.

Reviewer: ```suggestion
Notably, the `REGISTRY_AFFINITY_CUSTOM` variable overrides the default pod anti-affinity, and `REGISTRY_HPA_AUTO_SIZE` automatically adjusts the minimum and maximum replicas for the registry based on the number of nodes in the cluster. If you prefer to manually set the minimum and maximum replicas, you can use `REGISTRY_HPA_MIN` and `REGISTRY_HPA_MAX` to specify the desired values.
```
